text/0000-target-bundles.md

```diff
@@ -0,0 +1,245 @@
+    // configuration is painful.
+    is_like_osx: false,
+
+    // is_like_solaris is handled by the extra gc-sections key
+    // is_like_msvc is handled by the extra metadata-section, linker-kind, eh-method,
+    // dll-storage-attrs, debug-info keys
+    // is_like_windows is handled by the extra system-abi and c-abi-kind keys
+    // is_like_android is handled by the extra debug-info key
+}
+```
+
+## Distribution of targets
+
+Currently every built-in target is distributed along with the rustc compiler. This is suboptimal,
+because in majority of cases users are interested in targets which rustc can compile for, rather
+than the built-in targets rustc knows about, therefore distributing targets built-in into rustc is
+providing no benefits.
+
+Native libraries and linkers aside, it is obvious there’s little sense in distributing targets
+separately from the standard libraries for them. This RFC proposes to change the distribution so
+the rust-std and rust (not rustc or rust-docs) packages begin including the JSON target description
+for the target which the standard library targets. The target JSONs would get installed in
+`$sysroot/share/rustc/targets/` or a similar directory, and the `RUST_TARGET_PATH` environment
+variable would be adjusted to include this directory by default.
+
+This scheme allows users to easily produce and distribute custom standard library and target
+combinations, thus removing the need to land every single target as a built-in to rustc. Moreover,
+under this scheme, if rustc reports knowing about a target, it is very likely it will be able to
+compile for it as well, instead of reporting a confusing
+
+> error: can't find crate for `std` [E0463]
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This RFC does not attempt to solve issues surrounding linker invocation and native library
+discovery, especially during cross-compilation. Instead the code built-in into rustc is still relied
+onto to deal with these problems. On the other hand, it is likely that there isn’t much more
+necessary than rustc knowing how to invoke the few linkers it already knows about in order to cover
+the great majority of use cases.
+
+The users will be able to build the standard library-target bundles, but only for nightly versions
+of rustc compilers, because of the number of unstable features necessary to build a libstd. On the
+upside, it should become as easy as `cd src/libstd && cargo build --release
+--target=/path/to/custom-target.json && build_bundle`. If there’s a desire to make “bundles” work
+with stable rustc, the target would still be submitted upstream.
+
+`#![no_core]` users will have to download the rust-std even if they have no use for libraries in
+there.
+
+Proposed change to add a key for each option, instead of having an umbrella `is_like_*` keys, will
+result in big increase of such options. All of these are optional and should have sane default
+values, though.
+
+# Alternatives
+[alternatives]: #alternatives
```

Comment:
Seems like an alternative (in some sense) would be to have a `rustc --print target-spec` option that spit out more info.